src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
+                if (decrementedValue == 0)
+                {
+                    instance.Dispose();
+                }
+#else
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    Debug.Assert(currentValue > 0, "Dispose should have protected itself against races.");
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue - 1, currentValue) == currentValue)
+                    {
+                        if (currentValue == 1)
+                        {
+                            // Reference count hit 0 for this call
+                            instance.Dispose();
+                        }
+
+                        break;
+                    }
+                }
+#endif
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<ReferenceCountedDisposable<T>> _instance;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    var instance = reference._instance;
+                    Interlocked.MemoryBarrier();
+                    var referenceCount = Volatile.Read(ref reference._referenceCount);
+                    if (referenceCount == null)
+                    {
+                        // The specified reference is already not valid.
+                        return;
+                    }
+
+                    if (referenceCount.Value == 0)
+                    {
+                        // We were able to read the reference, but it's already disposed.
+                        return;
+                    }
+
+                    var innerReference = new ReferenceCountedDisposable<T>(instance, referenceCount);
+                    _instance = new WeakReference<ReferenceCountedDisposable<T>>(innerReference);
+                }
+
+                /// <summary>
+                /// Increments the reference count for the disposable object, and returns a new disposable reference to
+                /// it.
+                /// </summary>
+                /// <remarks>
+                /// <para>Unlike <see cref="ReferenceCountedDisposable{T}.TryAddReference"/>, this method is capable of
```

Comment:
:memo: Originally I had `TryGetTarget`, but it wasn't clear that the result needed to be disposed by the caller. While the details of these two methods is slightly different (one can return null while the other is still able to add a reference), they do share an important characteristic in the meaning of the return value is the same for the caller.